iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,429 @@
+    constexpr uint64_t size() const noexcept
+    {
+        return m_size;
+    }
+
+    span_storage(const span_storage& other) noexcept = default;
+    span_storage& operator=(const span_storage& other) noexcept = default;
+
+    span_storage(span_storage&& other) noexcept
+    {
+        *this = std::move(other);
+    }
+    span_storage& operator=(span_storage&& other) noexcept
+    {
+        if (this != &other)
+        {
+            m_size = other.m_size;
+            other.m_size = 0;
+        }
+        return *this;
+    }
+    ~span_storage() noexcept = default;
+
+  private:
+    uint64_t m_size{0};
+};
+
+template <typename T>
+constexpr T* to_address(T* p) noexcept
+{
+    static_assert(!std::is_function<T>::value, "Error: T must not be a function type.");
+    return p;
+}
+
+template <typename Ptr>
+constexpr auto to_address(const Ptr& p) noexcept -> decltype(std::pointer_traits<Ptr>::to_address(p))
+{
+    return std::pointer_traits<Ptr>::to_address(p);
+}
+
+template <typename Ptr, typename... None>
+constexpr auto to_address(const Ptr& p, None...) noexcept
+{
+    return to_address(p.operator->());
+}
+
+} // namespace detail
+
+/// @brief  C++11 compatible span implementation. The class template span describes an object that can refer to a
+///         contiguous sequence of objects with the first element of the sequence at position zero. A span can either
+///         have a static extent, in which case the number of elements in the sequence is known at compile-time and
+///         encoded in the type, or a dynamic extent. If a span has dynamic extent, a typical implementation holds two
+///         members: a pointer to T and a size. A span with static extent may have only one member: a pointer to T.
+///         http://eel.is/c++draft/views contains the latest C++20 draft
+/// @tparam T - element type; must be a complete object type that is not an abstract class type
+/// @tparam Extent - the number of elements in the sequence, or std::dynamic_extent if dynamic
+template <typename T, uint64_t Extent>
+class span : public detail::span_storage<Extent>
+{
+  private:
+    using span_storage_t = detail::span_storage<Extent>;
+
+  public:
+    using element_type = T;
+    using value_type = std::remove_cv_t<T>;
+    using uint64_type = uint64_t;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T*;
+    using reference = T&;
+    using iterator = span_iterator<T>;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    static constexpr uint64_t extent = Extent;
+
+    // constructors, copy, assignment, and destructor
+
+    constexpr span() noexcept = delete;
+
+    /// @brief Constructs a span that is a view over the range [first, first + count);
+    /// @tparam It Type of the iterator
+    /// @param first iterator of where to start creating the span
+    /// @param count number of elements counting from first
+    template <typename It>
+    constexpr span(It first, uint64_t count) noexcept;
```

Comment:
Nitpicking. In other occasions  the full name is used, i.e. here `Iterator`. No strong feelings about this though

Reply:
Which occasions do you mean?